pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,396 @@
+// value.
+func retrieveTableIDFromIfIndex(ifindex int, getter interfaceDB) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, fmt.Errorf("failed to find link by index: %w", err)
+	}
+
+	mac := link.Attrs().HardwareAddr.String()
+	ifaceNum, err := getter.GetInterfaceNumberByMAC(mac)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface number by MAC %q: %w", mac, err)
+	}
+
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
+}
+
+// onUpgradeInstallRulesAndRoutes migrates the ENI datapath for the upgrade
+// case. It accepts the old table ID for the existing rules to migrate. `new`
+// is the new table ID that the rules and routes should be referring to.
+func onUpgradeInstallRulesAndRoutes(
+	rules []netlink.Rule,
+	routes []netlink.Route,
+	oldTable, newTable int,
+) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	if revert, err := copyRoutes(routes, oldTable, newTable); err != nil {
+		return revertStack, err
+	} else {
+		revertStack.Extend(revert)
+	}
+
+	if revert, err := createNewRules(
+		rules,
+		linux_defaults.RulePriorityEgress,
+		oldTable,
+		linux_defaults.RulePriorityEgressv2,
+		newTable,
+	); err != nil {
+		return revertStack, err
+	} else {
+		revertStack.Extend(revert)
+	}
+
+	return revertStack, nil
+}
+
+// onUpgradeDeleteRulesAndRoutes removes the old ENI datapath for the upgrade
+// case. The old rules can be found by looking for the old table ID and a
+// specific priority (`prio`). The old routes can be found by looking for the
+// old table ID.
+func onUpgradeDeleteRulesAndRoutes(
+	rules []netlink.Rule,
+	routes []netlink.Route,
+	oldTable, prio int,
+) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	if revert, err := deleteOldRules(rules, prio, oldTable); err != nil {
+		return revert, err
+	} else {
+		revertStack.Extend(revert)
+	}
+
+	if revert, err := deleteOldRoutes(oldTable); err != nil {
+		return revert, err
+	} else {
+		revertStack.Extend(revert)
+	}
+
+	return revertStack, nil
+}
+
+func copyRoutes(routes []netlink.Route, from, to int) (revert.RevertStack, error) {
+	var revertStack revert.RevertStack
+
+	// In ENI mode, we only expect two rules:
+	//   1) Link scoped route with a gateway IP
+	//   2) Default route via gateway IP
+	// We must upsert the link scoped route first which will make the kernel
+	// aware of the gateway IP. After that, then we will upsert the "default
+	// via gatewayIP" route. If we didn't do this, then the kernel would
+	// complain with "Error: Nexthop has invalid gateway." with an errno of
+	// ENETUNREACH.
+	for _, r := range routes {
+		if r.Scope == netlink.SCOPE_LINK {
+			route := r
+			route.Table = to
+			if err := netlink.RouteReplace(&route); err != nil {
+				return revertStack, fmt.Errorf("unable to replace link scoped route under table ID: %w", err)
+			}
+
+			revertStack.Push(func() error {
+				return netlink.RouteDel(&route)
+			})
+
+			break
```

Comment:
I think the break here is kinda negligible. If the route table matches our expectations, then breaking early doesn't make much difference to performance. We're talking maybe stop iterating at 1/2 routes. If a user ended up doing something more interesting with the routes behind the scenes that leads there to be multiple link-scoped routes then this would skip some arbitrary subset of such routes. So in the common case it doesn't make a difference, but in the more complicated case it seems like it would be potentially inconsistent. :shrug:

Reply:
Good point, will remove.